tesseract/messaging/evm/src/provider.rs

```diff
@@ -903,6 +911,303 @@ impl IsmpProvider for EvmClient {
+	}
+}
+
+#[async_trait::async_trait]
+impl IsmpProvider for TendermintEvmClient {
+	async fn query_consensus_state(
+		&self,
+		at: Option<u64>,
+		id: ConsensusStateId,
+	) -> Result<Vec<u8>, Error> {
+		self.inner.query_consensus_state(at, id).await
+	}
+
+	async fn query_latest_height(&self, id: StateMachineId) -> Result<u32, Error> {
+		self.inner.query_latest_height(id).await
+	}
+
+	async fn query_finalized_height(&self) -> Result<u64, Error> {
+		self.inner.query_finalized_height().await
+	}
+
+	async fn query_state_machine_commitment(
+		&self,
+		height: StateMachineHeight,
+	) -> Result<StateCommitment, Error> {
+		self.inner.query_state_machine_commitment(height).await
+	}
+
+	async fn query_state_machine_update_time(
+		&self,
+		height: StateMachineHeight,
+	) -> Result<Duration, Error> {
+		self.inner.query_state_machine_update_time(height).await
+	}
+
+	async fn query_challenge_period(&self, id: StateMachineId) -> Result<Duration, Error> {
+		self.inner.query_challenge_period(id).await
+	}
+
+	async fn query_timestamp(&self) -> Result<Duration, Error> {
+		self.inner.query_timestamp().await
+	}
+
+	async fn query_requests_proof(
+		&self,
+		at: u64,
+		keys: Vec<Query>,
+		_counterparty: StateMachine,
+	) -> Result<Vec<u8>, Error> {
+		let mut proofs: Vec<EvmKVProof> = Vec::new();
+		for q in keys.into_iter() {
+			let contract_addr: [u8; 20] = self.inner.config.ismp_host.0;
+			let slot_hash =
+				super::derive_map_key(q.commitment.0.to_vec(), REQUEST_COMMITMENTS_SLOT);
+			let key_bytes = self.keys.storage_key(&contract_addr, slot_hash.0);
+			let resp = self
+				.prover
+				.abci_query_key(&self.store_key, key_bytes, at)
+				.await
+				.map_err(|e| anyhow::anyhow!("abci_query_key error: {e:?}"))?;
+			let proof_bytes = proof_ops_to_commitment_proof_bytes(resp.proof)?;
+			proofs.push(EvmKVProof { value: resp.value, proof: proof_bytes });
+		}
+		Ok(proofs.encode())
+	}
+
+	async fn query_responses_proof(
+		&self,
+		at: u64,
+		keys: Vec<Query>,
+		_counterparty: StateMachine,
+	) -> Result<Vec<u8>, Error> {
+		let mut proofs: Vec<EvmKVProof> = Vec::new();
+		for q in keys.into_iter() {
+			let contract_addr: [u8; 20] = self.inner.config.ismp_host.0;
+			let slot_hash =
+				super::derive_map_key(q.commitment.0.to_vec(), RESPONSE_COMMITMENTS_SLOT);
+			let key_bytes = self.keys.storage_key(&contract_addr, slot_hash.0);
+			let resp = self
+				.prover
+				.abci_query_key(&self.store_key, key_bytes, at)
+				.await
+				.map_err(|e| anyhow::anyhow!("abci_query_key error: {e:?}"))?;
+			let proof_bytes = proof_ops_to_commitment_proof_bytes(resp.proof)?;
+			proofs.push(EvmKVProof { value: resp.value, proof: proof_bytes });
+		}
+		Ok(proofs.encode())
+	}
+
+	async fn query_state_proof(
+		&self,
+		at: u64,
+		keys: StateProofQueryType,
+	) -> Result<Vec<u8>, Error> {
+		let mut proofs: Vec<EvmKVProof> = Vec::new();
+		match keys {
+			StateProofQueryType::Ismp(keys) | StateProofQueryType::Arbitrary(keys) => {
```

Comment:
Abritrary keys have length 52 bytes, where the first 20 bytes is the contract address